### PR TITLE
 Refactor function spec change detection

### LINKED ISSUE #50: The operator is consuming massive resources and crashing the Kubernetes API server
I have upgraded my openfaas-operator to 0.8.8 in order to fix #47  (Thanks for the quick fix!)
Unfortunately I am facing the similar problem.

## Expected Behaviour

Function deployments and services are updated without any error.

## Current Behaviour

After upgraded, the pods / deployments were getting killed and recreated again and again like in #47.

Then I dig into the log of the operator, and got the following

```bash
I0921 19:53:29.325471       1 deployment.go:241] Function my-func annotations have changed
I0921 19:53:29.325505       1 controller.go:316] Updating deployment for 'my-func'
E0921 19:53:30.406757       1 controller.go:328] Updating deployment for 'my-func' failed: Deployment.apps "my-func" is invalid: [metadata.annotations: Too long: must have at most 262144 characters, spec.template.annotations: Too long: must have at most 262144 characters]
E0921 19:53:32.305706       1 controller.go:339] Updating service for 'my-func' failed: Service "my-func" is invalid: metadata.annotations: Too long: must have at most 262144 characters
...
```

Similar to #47, the operator was looping these lines all the time, with a short interval.

After a while (like ~5 minutes), my monitor reported that 2 of my master nodes had crashed.

I checked the kubernetes-dashboard, it was reporting the deployment `gateway`, which has gateway and operator inside, was using ~8GB memory, I forgot the CPU usage, but it was not low.

Later I reverted the upgrade and things went back to normal.

## Possible Solution

I guess the operator was in a dead loop attempting to update the deployment and the service, but it couldn't make it because of the `*.metadata.annotations: Too long` problem.

## Steps to Reproduce (for bugs)

1. On an older version of `openfass-operator`, `0.8.6`, patched with the code snippet mentioned in #47
2. Deploy the following function

    registry-secret.yml
    ```yaml
    apiVersion: v1
    kind: Secret
    metadata:
    name: registry-secret
    namespace: faas
    type: kubernetes.io/dockercfg
    data:
    .dockercfg: xxxxxxx
    ```

    my-func-secret.yml
    ```yaml
    apiVersion: v1
    kind: Secret
    metadata:
    name: my-func-secret
    namespace: faas
    type: Opaque
    data:
    DB_HOST: xxx
    DB_NAME: xxx
    DB_PORT: xxx
    DB_USER: xxx
    DB_PASS: xxx
    ```

    stack.yml
    ```yaml
    provider:
    name: faas
    gateway: http://gateway

    functions:
    my-func:
        lang: go
        handler: ./function
        image: private.registry/my-func:0.0.1 
        environment:
        ENV_1: "1"
        combine_output: false
        secrets:
        - registry-secret
        - my-func-secret
    ```

    Please note that these are mocks, I am not allowed to share the real ones, sorry about that.
    Some stats for my real YAMLs
    * Length of the function name - `26 characters`
    * Length of the image name - `125 characters`
    * Length of the namespace - `8 characters`

3. Confirm the function is deployed successfully and working without error, test with OpenFaaS Portal
4. Edit the deployment with `kubectl edit`, change the image tag of the `openfaas-operator` to `0.8.8`
5. Disaster happens

## Context

I am trying to deploy a function which its image is in a private registry.

I provided the secret for the private registry via the fuction YAML.

It didn't work because the secrets part of the deployment comparison was not working probably, the deployments were always recognized changed, and the operator was always trying to update the deployment.

Then I raised #47, and 0.8.8 has been released to fix the problem.

However I still cannot make it working because of the `*.metadata.annotations: Too long` problem.

## Your Environment

* Docker version `docker version` (e.g. Docker 17.0.05 ):
```
17.09.0-ce
```

* What version and distriubtion of Kubernetes are you using? `kubectl version`
```bash
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.3", GitCommit:"a4529464e4629c21224b3d52edfe0ea91b072862", GitTreeState:"clean", BuildDate:"2018-09-10T11:44:36Z", GoVersion:"go1.11", Compiler:"gc", Platform:"darwin/amd64"}

Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.5", GitCommit:"32ac1c9073b132b8ba18aa830f46b77dcceb0723", GitTreeState:"clean", BuildDate:"2018-06-21T11:34:22Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```

* Operating System and version (e.g. Linux, Windows, MacOS):
```
Ubuntu 16.04.4 LTS
```

### PR DESCRIPTION
**Pull request by stefanprodan**
## Description

- use go-cmp to reliably detect changes in function spec
- log function spec diff before update
- fix deployment annotation append bug by copying the values from function spec
- set the default API namespace the same as the controller default
- change K8s event logger verbosity level to prevent log filling up when a function goes into crash loop, when debugging the event log can be enabled using the -v=3 flag

Fix  #50

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?

- [ ] I have run the [certifier](https://github.com/openfaas/certifier)

Tested with `stefanprodan/openfaas-operator:fix50v2` on GKE 1.10 by applying changes to every function spec field.

Function definition:

```yaml
apiVersion: openfaas.com/v1alpha2
kind: Function
metadata:
  name: env
spec:
  name: env
  image: stefanprodan/of-env
  annotations:
    test.min: "1"
    test.max: "2"
  labels:
    com.openfaas.scale.min: "2"
    com.openfaas.scale.max: "5"
  environment:
    test.min: "1"
    test.max: "2"
  secrets:
    - faas-token
    - faas-key
  limits:
    cpu: "100m"
    memory: "128Mi"
  requests:
    cpu: "10m"
    memory: "64Mi"
  constraints:
    - "cloud.google.com/gke-nodepool=default-pool"
```

**Tests**

Image change:

```
I0922 15:10:45.008390    4953 deployment.go:229] Change detected for env diff
{v1alpha2.FunctionSpec}.Image:
        -: "stefanprodan/of-env"
        +: "stefanprodan/of-env:1.1.0"
```

Annotation value change:

```
I0922 15:11:44.558130    4953 deployment.go:229] Change detected for env diff
(*{v1alpha2.FunctionSpec}.Annotations)["test.max"]:
        -: "2"
        +: "3"
```

Annotation key added:

```
I0922 15:13:30.037083    4953 deployment.go:229] Change detected for env diff
(*{v1alpha2.FunctionSpec}.Annotations)["test.add"]:
        -: <non-existent>
        +: "test"
```

Annotation key removed:

```
I0922 15:14:10.565108    4953 deployment.go:229] Change detected for env diff
(*{v1alpha2.FunctionSpec}.Annotations)["test.add"]:
        -: "test"
        +: <non-existent>
```

Secret rename:

```
I0922 15:14:59.623583    4953 deployment.go:229] Change detected for env diff
{v1alpha2.FunctionSpec}.Secrets[1]:
        -: "faas-key"
        +: "faas-key2"
```

Label change:

```
I0922 15:16:42.510263    4953 deployment.go:229] Change detected for env diff
(*{v1alpha2.FunctionSpec}.Labels)["com.openfaas.scale.min"]:
        -: "2"
        +: "1"
```

**Note**: annotations, labels, limit, requests and environment have the same type, all tested show the same behaviour. Same with secrets and constraints.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: pkg/controller/deployment.go [hunk 1]
```diff
@@ -199,121 +209,29 @@ func makeNodeSelector(constraints []string) map[string]string {
...
 		// is a new deployment or is an old deployment that is missing the annotation
 		return true
 	}
 
-	previousFunction := &faasv1.FunctionSpec{}
-	err := json.Unmarshal([]byte(previousFunctionSpec), previousFunction)
+	prevFnSpec := &faasv1.FunctionSpec{}
+	err := json.Unmarshal([]byte(prevFnSpecJson), prevFnSpec)
 	if err != nil {
-		glog.V(2).Infof("Failed to parse previous function spec: %s", err.Error())
```
Comment (alexellis) [inline pkg/controller/deployment.go]: If all this code is deleted (replaced by go-cmp) then we'll going to do some regression testing on each of the items that can cause a redeployment one by one and in various combinations.

File: pkg/server/server.go [hunk 2]
```diff
@@ -23,7 +23,7 @@ const defaultWriteTimeout = 8
 
 // Start starts HTTP Server for API
 func Start(client clientset.Interface, kube kubernetes.Interface) {
-	functionNamespace := "default"
+	functionNamespace := "openfaas-fn"
```
Comment (alexellis) [inline pkg/server/server.go]: Seems a valid change, but is this part of the fix?

File: pkg/controller/deployment.go [hunk 3]
```diff
@@ -148,18 +149,27 @@ func makeLabels(function *faasv1.Function) map[string]string {
...
-		annotations = map[string]string{}
+		for k, v := range *function.Spec.Annotations {
+			annotations[k] = v
+		}
 	}
 
-	annotations["prometheus.io.scrape"] = "false"
+	// save function spec in deployment annotations
+	// used to detect changes in function spec
+	specJSON, err := json.Marshal(function.Spec)
```
Comment (alexellis) [inline pkg/controller/deployment.go]: Can you confirm that we can no longer get the bug with the infinite annotation? Would be good to see a test showing that the bug is present in the old code and then see it fixed in the new code. 

`makeAnnotations` should be easy to add tests for.

File: pkg/controller/deployment.go [hunk 4]
```diff
@@ -199,121 +209,29 @@ func makeNodeSelector(constraints []string) map[string]string {
 
 // deploymentNeedsUpdate determines if the function spec is different from the deployment spec
```
Comment (alexellis) [inline pkg/controller/deployment.go]: Would be ideal to see a test on `deploymentNeedsUpdate`

**Comment by alexellis** [comment]
This will need some exploration testing with adding / removing and updating each field covered by the spec.

File: pkg/controller/deployment.go (hunk 3 above)
Comment (stefanprodan) [reply pkg/controller/deployment.go]: I've manually tested changes to secrets, annotations and labels (keys and values), image, environment, limits and requests. The JSON annotation was updated not appended and all changes were detected correctly.

**Comment by stefanprodan** [comment]
@alexellis I've tested all the function spec fields: secrets (add, delete, rename), annotations and labels (keys and values), image, environment, limits, requests and constraints. I've also added tests for make annotations func. I've updated the PR description with some examples.

File: pkg/controller/annotations_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
+package controller
+
+import (
+	"testing"
+
+	faasv1 "github.com/openfaas-incubator/openfaas-operator/pkg/apis/openfaas/v1alpha2"
+)
+
+func Test_makeAnnotations_NoKeys(t *testing.T) {
```
Comment (alexellis) [inline pkg/controller/annotations_test.go]: This I s basically the tests from my gist? Any other scenarios we should cover? cc @rgee0

Comment (johnmccabe) [reply pkg/controller/annotations_test.go]: Setting `prometheus.io.scrape` up front in the `annotationVal` as a `string`, `bool`, number, `null`

**Pull request merged**